File: FE/canonicalize_large_args.py
"""Canonicalize Large Args (and Results)

The goal of this step is to simplify code generation.
We want to avoid having to deal with values that do not fit into registers.
Those values are primarily rec and sum-types. Slices would also fall into this
category but we convert them to rec in the prior step.

The only place where we allow values  that do not fit into register
are let statements.

Effects of this step:
* Rewrite arguments and results which cannot be passed in registers
  as pointers.
* Large results will become an extra pointer parameter appended to the end
  of the parameter list.

"""

from typing import Optional, Any

from FE import cwast
from FE import type_corpus
from FE import typify
from FE import eval
from FE import canonicalize

from IR import opcode_tab as o

############################################################
# Convert large parameter into pointer to object allocated
# in the caller
############################################################


def MakeAndRegisterLargeArgReplacements(tc: type_corpus.TypeCorpus):
    tc.ClearReplacementInfo()
    for fun_sig in tc.topo_order[:]:
        if not fun_sig.is_fun():
            continue
        change = False
        params: list[cwast.CanonType] = []
        for p in fun_sig.parameter_types():
            if p.ir_regs == o.DK.MEM:
                params.append(tc.InsertPtrType(False, p))
                change = True
            else:
                params.append(p)
        result = fun_sig.result_type()
        if result.ir_regs == o.DK.MEM:
            change = True
            params.append(tc.InsertPtrType(True, result))
            result = tc.get_void_canon_type()
        if change:
            fun_sig.LinkReplacementType(tc.InsertFunType(params, result))


def _FixupFunctionSignature(fun: cwast.DefFun,  old_sig: cwast.CanonType, new_sig: cwast.CanonType,
                            tc: type_corpus.TypeCorpus) -> Optional[cwast.FunParam]:
    result_param = None
    typify.NodeChangeType(fun, new_sig)
    result_changes = old_sig.result_type() != new_sig.result_type()
    if result_changes:
        assert len(new_sig.parameter_types()) == 1 + \
            len(old_sig.parameter_types())
        assert new_sig.result_type().is_void()
        typify.NodeChangeType(fun.result, tc.get_void_canon_type())
        #
        sl = fun.x_srcloc
        at = cwast.TypeAuto(sl, new_sig.parameter_types()[-1])
        result_param = cwast.FunParam(cwast.NAME.Make(
            "large_result"), at, x_srcloc=sl, x_type=at.x_type, res_ref=True)
        fun.params.append(result_param)
        assert isinstance(fun.result, cwast.TypeAuto)

    # note: new_sig may contain an extra param at the end
    for p, old, new in zip(fun.params, old_sig.parameter_types(), new_sig.parameter_types()):
        if old != new:
            assert isinstance(p.type, cwast.TypeAuto)
            typify.NodeChangeType(p, new)
            typify.NodeChangeType(p.type, new)
            p.arg_ref = True
    return result_param


def FunRewriteLargeArgsParameter(fun: cwast.DefFun, new_fun_ct: cwast.CanonType,
                                 tc: type_corpus.TypeCorpus):
    old_fun_ct: cwast.CanonType = fun.x_type

    result_param = _FixupFunctionSignature(fun, old_fun_ct,  new_fun_ct, tc)

    # print([k.name for k, v in changing_params.items()], result_changes)

    def replacer(node) -> Optional[Any]:

        if isinstance(node, cwast.Id) and isinstance(node.x_symbol, cwast.FunParam) and node.x_symbol.arg_ref:
            new_node = cwast.ExprDeref(
                node, x_srcloc=node.x_srcloc, x_type=node.x_type)
            typify.NodeChangeType(node, node.x_symbol.x_type)
            return new_node

        elif isinstance(node, cwast.StmtReturn) and node.x_target == fun and result_param:
            result_ct: cwast.CanonType = result_param.type.x_type
            assert result_ct.is_ptr()
            sl = node.x_srcloc
            lhs = cwast.ExprDeref(
                cwast.Id(result_param.name, None, x_srcloc=sl,
                         x_type=result_ct, x_symbol=result_param),
                x_srcloc=sl, x_type=result_ct.underlying_type())
            assign = cwast.StmtAssignment(
                lhs, node.expr_ret, x_srcloc=node.x_srcloc)
            node.expr_ret = cwast.ValVoid(
                x_srcloc=sl, x_type=tc.get_void_canon_type(), x_eval=eval.VAL_VOID)
            return [assign, node]
        return None

    cwast.MaybeReplaceAstRecursivelyPost(fun, replacer)


def MakeDefAndAddrOfForParam(name, init, ct_old: cwast.CanonType,
                             ct_new, mut: bool) -> tuple[cwast.DefVar,
                                                         cwast.ExprAddrOf]:
    sl = init.x_srcloc
    at = cwast.TypeAuto(x_srcloc=sl, x_type=ct_old)
    def_var = cwast.DefVar(name, at, init, x_srcloc=sl,
                           x_type=ct_old, ref=True, mut=mut)
    sym = canonicalize.IdNodeFromDef(def_var, sl)
    addr_of = cwast.ExprAddrOf(sym, x_srcloc=sl, x_type=ct_new)
    return def_var, addr_of


def FixupCall(call: cwast.ExprCall, orig_fun_ct: cwast.CanonType,
              new_fun_ct: cwast.CanonType, tc: type_corpus.TypeCorpus):
    sl = call.x_srcloc

    typify.NodeChangeType(call.callee, new_fun_ct)
    expr_body: list[Any] = []
    expr = cwast.ExprStmt(
        expr_body, x_srcloc=call.x_srcloc, x_type=call.x_type)
    # note: new_sig might be longer if the result type was changed
    for n, (old_ct, new_ct) in enumerate(zip(orig_fun_ct.parameter_types(),
                                             new_fun_ct.parameter_types())):
        if old_ct == new_ct:
            continue
        # Note: the type of this new variable is NOT call.args[n].x_type
        # BUT old since we still rely on some basic implicit conversion.
        # E.g. from mut to non-mut.
        new_def, addr_of = MakeDefAndAddrOfForParam(cwast.NAME.Make(f"arg{n}"),
                                                    call.args[n], old_ct, new_ct, mut=False)
        expr_body.append(new_def)
        call.args[n] = addr_of
    if len(orig_fun_ct.parameter_types()) != len(new_fun_ct.parameter_types()):
        old_ct = orig_fun_ct.result_type()
        new_ct = new_fun_ct.parameter_types()[-1]
        undef_init = cwast.ValUndef(x_srcloc=sl, x_eval=eval.VAL_UNDEF)
        # the result is not a argument
        new_def, addr_of = MakeDefAndAddrOfForParam(cwast.NAME.Make("result"),
                                                    undef_init, old_ct, new_ct, mut=True)
        call.args.append(addr_of)
        typify.NodeChangeType(call, tc.get_void_canon_type())
        expr_body.append(new_def)
        expr_body.append(cwast.StmtExpr(call, x_srcloc=sl))
        expr_body.append(cwast.StmtReturn(
            canonicalize.IdNodeFromDef(new_def, sl), x_srcloc=call.x_srcloc, x_target=expr))
    else:
        expr_body.append(cwast.StmtReturn(
            call, x_srcloc=sl, x_target=expr))
    return expr


def FunRewriteLargeArgsCallsites(fun: cwast.DefFun, tc: type_corpus.TypeCorpus):
    """Assuming the callee signature was changed like so
          foo(a: rec A, b: rec B, c: rec C)
       To
          foo(a: ptr(rec A), b: ptr(rec B), c: ptr(rec C))


       The call site changes from;
           foo(a, b, c)
       To
           aa = a
           bb = b
           cc = c
           foo(&aa, &bb, &cc)

       We have to play it safe and materialize aa, bb, cc under certain conditions:
       * a, b, c are expression
         we cannot take address of an expression so we have to make copy first.
       * a, b, c are mutable storage locations
         we cannot be sure if mutable aliases to a, b, c are accessed by foo
         of functions called by foo, so we have to make a copy with value just before the call.
    """
    def replacer(call, _parent) -> Optional[Any]:
        if not isinstance(call, cwast.ExprCall):
            return None
        orig_fun_ct: cwast.CanonType = call.callee.x_type
        new_fun_ct: cwast.CanonType = orig_fun_ct.replacement_type
        if new_fun_ct is None:
            return None
        return FixupCall(call, orig_fun_ct, new_fun_ct, tc)

    cwast.MaybeReplaceAstRecursivelyWithParentPost(fun, replacer)

File: FE/canonicalize_large_args.cc


#include "FE/canonicalize_large_args.h"

#include <array>

#include "FE/canonicalize.h"
#include "FE/cwast_gen.h"
#include "FE/eval.h"
#include "FE/typify.h"

namespace cwerg::fe {

void MakeAndRegisterLargeArgReplacements(TypeCorpus* tc) {
  tc->ClearReplacementInfo();

  for (CanonType orig_fun_ct : tc->InTopoOrder()) {
    if (!CanonType_is_fun(orig_fun_ct)) continue;

    bool change = false;
    const std::span<CanonType> orig_children_ct =
        CanonType_children(orig_fun_ct);

    std::vector<CanonType> new_children_ct;

    for (CanonType child :
         orig_children_ct.first(orig_children_ct.size() - 1)) {
      DK ir_regs = CanonType_ir_regs(child);
      if (ir_regs == DK::MEM) {
        new_children_ct.push_back(tc->InsertPtrType(false, child));
        change = true;
      } else {
        new_children_ct.push_back(child);
      }
    }

    CanonType new_result_ct = orig_children_ct.back();
    if (CanonType_ir_regs(new_result_ct) == DK::MEM) {
      new_children_ct.push_back(tc->InsertPtrType(true, new_result_ct));
      new_result_ct = tc->get_void_canon_type();
      change = true;
    }
    new_children_ct.push_back(new_result_ct);

    if (change) {
      CanonType new_fun_ct = tc->InsertFunType(new_children_ct);
      CanonTypeLinkReplacementType(orig_fun_ct, new_fun_ct);
    }
  }
}

Node FixupFunctionSignature(Node fun, CanonType old_sig, CanonType new_sig,
                            TypeCorpus* tc) {
  Node result_param = kNodeInvalid;
  NodeChangeType(fun, new_sig);
  const std::span<CanonType> old_children_ct = CanonType_children(old_sig);
  const std::span<CanonType> new_children_ct = CanonType_children(new_sig);
  // fixup large result
  if (old_children_ct.size() != new_children_ct.size()) {
    NodeChangeType(Node_result(fun), tc->get_void_canon_type());
    //
    const SrcLoc& sl = Node_srcloc(fun);
    Node at = MakeTypeAuto(new_children_ct[new_children_ct.size() - 2], sl);
    result_param = NodeNew(NT::FunParam);
    NodeInitFunParam(result_param, NameNew("large_result"), at,
                     Mask(BF::RES_REF), kStrInvalid, sl, Node_x_type(at));
    if (Node_params(fun).isnull()) {
      Node_params(fun) = result_param;
    } else {
      Node p = Node_params(fun);
      while (!Node_next(p).isnull()) {
        p = Node_next(p);
      }
      Node_next(p) = result_param;
    }
  }
  // other param
  Node param = Node_args(fun);
  for (int i = 0; i < old_children_ct.size() - 1; ++i) {
    CanonType old_ct = old_children_ct[i];
    CanonType new_ct = new_children_ct[i];
    if (old_ct != new_ct) {
      NodeChangeType(param, new_ct);
      NodeChangeType(Node_type(param), new_ct);
      Node_set_flag(param, BF::ARG_REF);
    }
    param = Node_next(param);
  }
  return result_param;
}

void FunRewriteLargeArgsParameter(Node fun, CanonType old_sig,
                                  CanonType new_sig, TypeCorpus* tc) {
  Node result_param = FixupFunctionSignature(fun, old_sig, new_sig, tc);
  auto replacer = [fun, result_param, tc](Node node, Node parent) -> Node {
    if (node.kind() == NT::Id) {
      Node sym = Node_x_symbol(node);
      if (sym.kind() == NT::FunParam && Node_has_flag(sym, BF::ARG_REF)) {
        Node out = NodeNew(NT::ExprDeref);
        NodeInitExprDeref(out, node, kStrInvalid, Node_srcloc(node),
                          Node_x_type(node));
        NodeChangeType(node, Node_x_type(sym));
        return out;
      }
    } else if (node.kind() == NT::StmtReturn && !result_param.isnull() &&
               Node_x_target(node) == fun) {
      NodeChain out;
      CanonType result_ct = Node_x_type(result_param);
      const SrcLoc& sl = Node_srcloc(node);
      Node id = NodeNew(NT::Id);
      NodeInitId(id, Node_name(result_param), kStrInvalid, sl,
                 result_param, result_ct);
      Node lhs = NodeNew(NT::ExprDeref);
      NodeInitExprDeref(lhs, id, kStrInvalid, sl,
                        CanonType_underlying_type(result_ct));
      Node assign = NodeNew(NT::StmtAssignment);
      NodeInitStmtAssignment(assign, lhs, Node_expr_ret(node), kStrInvalid, sl);
      Node val_void = NodeNew(NT::ValVoid);
      NodeInitValVoid(val_void, kStrInvalid, sl, tc->get_void_canon_type());
      Node_x_eval(val_void) = kConstVoid;
      Node_expr_ret(node) = val_void;
      out.Append(assign);
      out.Append(node);
      return out.First();
    }

    return node;
  };
  MaybeReplaceAstRecursivelyPost(fun, replacer, kNodeInvalid);
}

struct DefAndArg {
  Node def;
  Node arg;
};

DefAndArg MakeDefVarAndAddrOfForParam(Name name, Node init, CanonType ct_old,
                                      CanonType ct_new, bool mut) {
  const SrcLoc& sl = Node_srcloc(init);
  Node at = MakeTypeAuto(ct_old, sl);
  Node def = NodeNew(NT::DefVar);
  uint16_t flags = Mask(BF::REF) | (mut ? Mask(BF::MUT) : 0);
  NodeInitDefVar(def, name, at, init, flags, kStrInvalid, sl, ct_old);
  Node sym = IdNodeFromDef(def, sl);
  Node addr_of = NodeNew(NT::ExprAddrOf);
  NodeInitExprAddrOf(addr_of, sym, 0, kStrInvalid, sl, ct_new);
  return {def, addr_of};
}

Node FixupCall(Node call, CanonType orig_fun_ct, CanonType new_fun_ct,
               TypeCorpus* tc) {
  const SrcLoc& sl = Node_srcloc(call);
  Node_next(call) = kNodeInvalid;  // unlink call before consuming it later

  NodeChain expr_body;
  NodeChain new_args;
  Node out = NodeNew(NT::ExprStmt);
  NodeInitExprStmt(out, kNodeInvalid, kStrInvalid, sl, Node_x_type(call));
  //
  NodeChangeType(Node_callee(call), new_fun_ct);
  const std::span<CanonType> orig_children_ct = CanonType_children(orig_fun_ct);
  const std::span<CanonType> new_children_ct = CanonType_children(new_fun_ct);

  Node arg = Node_args(call);
  for (int i = 0; i < orig_children_ct.size() - 1; ++i) {
    Node the_arg = arg;
    arg = Node_next(arg);
    Node_next(the_arg) = kNodeInvalid;
    ASSERT(!the_arg.isnull(), "");
    CanonType old_ct = orig_children_ct[i];
    CanonType new_ct = new_children_ct[i];
    if (old_ct != new_ct) {
      auto def_and_arg = MakeDefVarAndAddrOfForParam(NameNew("arg", i), the_arg,
                                                     old_ct, new_ct, false);
      expr_body.Append(def_and_arg.def);
      the_arg = def_and_arg.arg;
    }
    new_args.Append(the_arg);
  }
  //
  if (orig_children_ct.size() != new_children_ct.size()) {
    Node undef_init = NodeNew(NT::ValUndef);
    NodeInitValUndef(undef_init, kStrInvalid, Node_srcloc(call));
    Node_x_eval(undef_init) = kConstUndef;
    auto [new_def, the_arg] = MakeDefVarAndAddrOfForParam(
        NameNew("result"), undef_init, orig_children_ct.back(),
        new_children_ct[new_children_ct.size() - 2], true);
    new_args.Append(the_arg);
    expr_body.Append(new_def);
    Node call_stmt = NodeNew(NT::StmtExpr);
    NodeChangeType(call, tc->get_void_canon_type());
    expr_body.Append(NodeInitStmtExpr(call_stmt, call, kStrInvalid, sl));
    Node ret_stmt = NodeNew(NT::StmtReturn);
    NodeInitStmtReturn(ret_stmt, IdNodeFromDef(new_def, sl), kStrInvalid, sl,
                       out);
    expr_body.Append(ret_stmt);
  } else {
    Node ret_stmt = NodeNew(NT::StmtReturn);
    NodeInitStmtReturn(ret_stmt, call, kStrInvalid, sl, out);
    expr_body.Append(ret_stmt);
  }
  Node_body(out) = expr_body.First();
  Node_args(call) = new_args.First();
  return out;
}

void FunRewriteLargeArgsCallsites(Node fun, TypeCorpus* tc) {
  auto replacer = [tc](Node call, Node parent) -> Node {
    if (call.kind() != NT::ExprCall) return call;
    CanonType orig_fun_ct = Node_x_type(Node_callee(call));
    CanonType new_fun_ct = tc->MaybeGetReplacementType(orig_fun_ct);
    if (new_fun_ct.isnull()) return call;
    return FixupCall(call, orig_fun_ct, new_fun_ct, tc);
  };
  MaybeReplaceAstRecursivelyPost(fun, replacer, kNodeInvalid);
}

}  // namespace cwerg::fe